microovn/api/types/ovsdb.go

```diff
@@ -0,0 +1,118 @@
+
+// UpgradeRequired method returns "true" if there's any discrepancy between currently running OVSDB schema version
+// and a schema version expected by the individual hosts in the cluster. It also returns "true", if any of the cluster
+// members had Error in their OvsdbSchemaNodeStatus as it may indicate either a host that needs an attention because
+// it's down, or it's an older version of the MicroOVN that does not support API for fetching expected schema version.
+//
+// If all values match and there was no error, this method returns "false".
+//
+// This method returns error if Nodes is an empty list.
+func (s *OvsdbSchemaClusterStatus) UpgradeRequired(ctx context.Context) (bool, error) {
+	if len(s.Nodes) == 0 {
+		return false, errors.New("can't perform 'upgrade check' on empty cluster")
+	}
+
+	versionSet := make(map[string]int)
+	nodeError := false
+	expectedVersion := "0.0"
+
+	// Iterate over all nodes and find unique expected versions as well as any errors
+	for _, node := range s.Nodes {
+		expectedVersion = node.ExpectedVersion
+		versionSet[node.ExpectedVersion]++
+		if node.Error {
+			nodeError = true
+		}
+	}
+
+	// return True if any errors were encountered
+	if nodeError {
+		return true, nil
+	}
+
+	// return True if there's more than one unique schema version expected
+	if len(versionSet) > 1 {
+		return true, nil
+	}
+
+	// return True if expected schema version does not match the currently running version
+	_, err := shared.RunCommandContext(
+		ctx,
+		"ovsdb-tool",
+		"compare-versions",
+		s.RunningVersion,
+		"==",
+		expectedVersion,
+	)
+
+	return err != nil, nil
+
+}
+
+// Format returns a string that verbosely describes contents of the OvsdbSchemaClusterStatus.
+//
+// If the cluster does not require update, a simple "OK" is returned. Otherwise, this method returns string
+// that describes currently running OVSDB schema version, schema versions expected by the individual cluster members,
+// and whether there were any errors when fetching the expected schema from other cluster members.
+func (s *OvsdbSchemaClusterStatus) Format(ctx context.Context, cli *client.Client) string {
+	upgradeRequired, err := s.UpgradeRequired(ctx)
+	if err != nil {
+		return fmt.Sprintf("%v\n", err)
+	}
+
+	if !upgradeRequired {
+		return "OK\n"
+	}
+
+	clusterMembers, _ := cli.GetClusterMembers(ctx)
+
+	addrToName := func(addr string) (string, error) {
+		for _, member := range clusterMembers {
+			if member.Address.Addr().String() == addr {
+				return member.Name, nil
+			}
+		}
+		return "", fmt.Errorf("cluster member with address '%s' not found", addr)
+	}
+
+	msg := "Upgrade or attention required!\n"
+	msg += fmt.Sprintf("Currently running schema: %s\n", s.RunningVersion)
+	msg += "Cluster report (expected schema versions):\n"
+	for _, node := range s.Nodes {
+		nodeName, err := addrToName(node.Host)
+		if err != nil {
+			nodeName = node.Host
+		}
+		msg += fmt.Sprintf("\t%s: ", nodeName)
+		if node.Error {
+			msg += "N/A\n"
```

Comment:
This message is printed in multiple situations, and it does not appear to be particularly helpful for the operator.

1) For example when only one node of the cluster has been upgraded, the non-upgraded nodes will be shown as `N/A` regardless of them having central components that actually do need to be upgraded or not, example:
```
$ sudo microovn status
MicroOVN deployment summary:
- micro-1 (10.5.0.2)
  Services: central, chassis, switch

- micro-2 (10.5.0.3)
  Services: central, chassis, switch

- micro-3 (10.5.0.4)
  Services: central, chassis, switch

- micro-4 (10.5.0.5)
  Services: chassis, switch

OVN Database summary:
OVN Southbound: Upgrade or attention required!
Currently running schema: 20.21.0
Cluster report (expected schema versions):
    micro-1: 20.33.0
    micro-3: N/A
    micro-2: N/A
    micro-4: N/A

OVN Northbound: Upgrade or attention required!
Currently running schema: 6.1.0
Cluster report (expected schema versions):
    micro-1: 7.3.0
    micro-2: N/A
    micro-3: N/A
    micro-4: N/A
```

2)  When all central nodes are upgraded and the chassis nodes are not upgraded, the schema conversion will (rightfully) be held back, but the status does not reveal any information to the end user that this is the reason for the upgrade not happening.

```
$ sudo microovn status
MicroOVN deployment summary:
- micro-1 (10.5.0.2)
  Services: central, chassis, switch

- micro-2 (10.5.0.3)
  Services: central, chassis, switch

- micro-3 (10.5.0.4)
  Services: central, chassis, switch

- micro-4 (10.5.0.5)
  Services: chassis, switch

OVN Database summary:
OVN Southbound: Upgrade or attention required!
Currently running schema: 20.21.0
Cluster report (expected schema versions):
    micro-1: 20.33.0
    micro-4: N/A
    micro-3: 20.33.0
    micro-2: 20.33.0

OVN Northbound: Upgrade or attention required!
Currently running schema: 6.1.0
Cluster report (expected schema versions):
    micro-1: 7.3.0
    micro-4: N/A
    micro-3: 7.3.0
    micro-2: 7.3.0
```

So for the chassis node could we perhaps print a hint about snap upgrade being required? Another possiblitiy is to use the word "supported" instead of "expected" for schema version, that way it would not be confusing when we print schema version for the "client" nodes?

Mock-up:
```
Cluster report (supported schema versions):
    micro-1: 7.3.0
    micro-4: Unknown - snap upgrade required
    micro-3: 7.3.0
    micro-2: 7.3.0

Cluster report (supported schema versions):
    micro-1: 7.3.0
    micro-4: 6.1.0 - snap upgrade required
    micro-3: 7.3.0
    micro-2: 7.3.0
```

Reply:
Thank you for the review and testing @fnordahl. For starters I want to clarify that every node, regardless of whether it runs central services or not, is expected to report "expected version" for the schema. However, we are facing kind of a one-off situation here where non-upgraded nodes do not have the API required for this feature to work. Therefore any non-upgraded nodes report `N/A`, in the future these nodes would report older schema version.
We can probably distinguish `404` response code and print more informative message in this scenario. That would leave the `N/A` value solely for unexpected errors during the expected schema version fetch.

So in the current scenario, where some hosts in the cluster do not support this feature you'd see something like:

OVN Database summary:
OVN Southbound: Upgrade or attention required!
Currently running schema: 20.21.0
Cluster report (expected schema versions):
    micro-1: 20.33.0
    micro-4: API not supported - MicroOVN requires upgrade
    micro-3: 20.33.0
    micro-2: 20.33.0